tests/30rooms/13anonymousaccess.pl

```diff
@@ -62,6 +62,163 @@
+
+   do => sub {
+      my ( $api_client, $user ) = @_;
+
+      my $anonymous_user;
+      my $room_id;
+
+      register_anonymous_user( $api_client )->then( sub {
+         ( $anonymous_user ) = @_;
+
+         matrix_create_and_join_room( [ $user ] )
+         ->then( sub {
+            ( $room_id ) = @_;
+
+            set_room_history_visibility( $user, $room_id, "world_readable" );
+         })->then( sub {
+            Future->needs_all(
+               delay( 0.05 )->then( sub {
+                  matrix_send_room_text_message( $user, $room_id, body => "mice" );
+               }),
+
+               do_request_json_for( $anonymous_user,
+                  method => "GET",
+                  uri    => "/api/v1/events",
+                  params => {
+                     limit => "2",
+                     dir   => "b",
+                  },
+               )->main::expect_http_400->then( sub {
+                  do_request_json_for( $anonymous_user,
+                     method => "GET",
+                     uri    => "/api/v1/events",
+                     params => {
+                        limit   => "2",
+                        dir     => "b",
+                        room_id => $room_id,
+                     },
+                  )
+               })->then( sub {
+                  my ( $body ) = @_;
+
+                  require_json_keys( $body, qw( chunk ) );
+                  $body->{chunk} >= 1 or die "Want at least one event";
+                  my $chunk = $body->{chunk}[0];
+                  require_json_keys( $chunk, qw( content ) );
+                  my $content = $chunk->{content};
+                  require_json_keys( $content, qw( body ) );
+                  $content->{body} eq "mice" or die "Want content body to be mice";
+
+                  Future->done( 1 );
+               }),
+            );
+         });
+      });
+   };
+
+test "Anonymous user doesn't get events before room made world_readable",
+   requires => [ qw( first_api_client ), local_user_preparer() ],
+
+   do => sub {
+      my ( $api_client, $user ) = @_;
+
+      my $anonymous_user;
+      my $room_id;
+
+      register_anonymous_user( $api_client )->then( sub {
+         ( $anonymous_user ) = @_;
+
+         matrix_create_and_join_room( [ $user ] )
+         ->then( sub {
+            ( $room_id ) = @_;
+
+            Future->needs_all(
+               delay( 0.05 )->then( sub {
+                  matrix_send_room_text_message( $user, $room_id, body => "private" )->then(sub {
+                     set_room_history_visibility( $user, $room_id, "world_readable" );
+                  })->then( sub {
+                     matrix_send_room_text_message( $user, $room_id, body => "public" );
+                  });
+               }),
+
+               check_events( $anonymous_user, $room_id )
+               ->then(sub {
+                  Future->done( 1 );
+               }, sub {
+                  check_events( $anonymous_user, $room_id );
```

Comment:
This is an interesting construction. I vaguely remember @leonerd saying that he had a way to repeatedly try something until it worked to work around things failing because the server was still processing something.